### PR TITLE
Enable user RTC memory

### PR DESCRIPTION
**Pull request by lask**

### CODE REVIEW

File: lib/esp32.toit [hunk 1]
```diff
@@ -100,3 +100,19 @@ image_config_ -> ByteArray:
 
 set_real_time_clock_ seconds/int ns/int -> none:
   #primitive.core.set_real_time_clock
+
+/**
+Sets the RTC user data to the given $data starting from the given $offset.
+
+The given $data must be within bounds. That is 0 <= $offset <= $offset + data.size < 2048.
+*/
+set_rtc_data data offset/int=0 -> none:
```
Comment (floitsch) [inline lib/esp32.toit:109]: Depending on how nice we want to make the RTC api, we could provide a singleton with `operator[]` ..., but as an underlying API this is probably good enough.

File: lib/esp32.toit [hunk 2]
```diff
@@ -100,3 +100,19 @@ image_config_ -> ByteArray:
 
 set_real_time_clock_ seconds/int ns/int -> none:
   #primitive.core.set_real_time_clock
+
+/**
+Sets the RTC user data to the given $data starting from the given $offset.
```
Comment (floitsch) [inline lib/esp32.toit:105]: Add some explanation what RTC is.

File: lib/esp32.toit [hunk 3]
```diff
@@ -100,3 +100,19 @@ image_config_ -> ByteArray:
...
+
+The given $data must be within bounds. That is 0 <= $offset <= $offset + data.size < 2048.
```
Comment (floitsch) [inline lib/esp32.toit:107]: I would add a constant `RTC_SIZE` and then reference it here:
```suggestion
The given $data must be within bounds. That is 0 <= $offset <= $offset + data.size < $RTC_SIZE.
```

File: src/rtc_memory_esp32.h [hunk 4]
```diff
@@ -24,6 +24,8 @@ namespace toit {
 // The RTC memory holds state that is preserved across reboots.
 class RtcMemory {
  public:
+  static const int RTC_USER_DATA_SIZE = 4096;
```
Comment (floitsch) [inline src/rtc_memory_esp32.h:27]: In the comment for the Toit functions it says 2K.

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/rtc_memory_esp32.h (hunk 4 above)
Comment (lask) [reply src/rtc_memory_esp32.h:27]: You asked about the amount of RTC memory we have available. It seems like we have a total of 8K available. We use a bit for boot count and related things, so we could in principle bump this further.

It should be no problem to increase it (maybe we shouldn't have a default value for the length in the toit-level primitive calls if we want to change it).

File: lib/esp32.toit (hunk 3 above)
Comment (lask) [reply lib/esp32.toit]: Called it `RTC_MEMORY_SIZE`.

File: lib/esp32.toit (hunk 2 above)
Comment (lask) [reply lib/esp32.toit:108]: Added advanced section with a bit of explanation.

File: lib/esp32.toit (hunk 1 above)
Comment (lask) [reply lib/esp32.toit:118]: We could definitely do that, but we should probably add a primitive specifically for that to reduce the overhead.

Comment (lask) [reply lib/esp32.toit:118]: Maybe the correct thing to do is to make a `ByteArray` with an external allocation that points to the RTC memory.

**Comment by lask** [comment]
I added an alternative implementation that uses external byte arrays: https://github.com/toitlang/toit/pull/368

**Comment by lask** [comment]
Abandoning this PR in favor of #368.

**Pull request closed without merging**